geode-wan/src/main/java/org/apache/geode/management/internal/cli/functions/WanCopyRegionFunction.java

```diff
@@ -0,0 +1,581 @@
+
+    if (regionName.startsWith(SEPARATOR)) {
+      regionName = regionName.substring(1);
+    }
+    if (regionName.equals("*") && senderId.equals("*") && isCancel) {
+      return cancelAllWanCopyRegion(context);
+    }
+
+    if (isCancel) {
+      return cancelWanCopyRegion(context, regionName, senderId);
+    }
+    final Cache cache = context.getCache();
+
+    final Region<?, ?> region = cache.getRegion(regionName);
+    if (region == null) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__REGION__NOT__FOUND, regionName));
+    }
+    GatewaySender sender = cache.getGatewaySender(senderId);
+    if (sender == null) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__SENDER__NOT__FOUND, senderId));
+    }
+    if (!region.getAttributes().getGatewaySenderIds().contains(sender.getId())) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__REGION__NOT__USING_SENDER, regionName,
+              senderId));
+    }
+    if (!sender.isRunning()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__SENDER__NOT__RUNNING, senderId));
+    }
+    if (!sender.isParallel() && !((InternalGatewaySender) sender).isPrimary()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__SENDER__SERIAL__AND__NOT__PRIMARY,
+              senderId));
+    }
+
+    try {
+      return executeWanCopyRegionFunctionInNewThread(context, region, regionName, sender, maxRate,
+          batchSize);
+    } catch (InterruptedException | CancellationException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.WAN_COPY_REGION__MSG__CANCELED__BEFORE__HAVING__COPIED);
+    } catch (ExecutionException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__EXECUTION__FAILED, e.getMessage()));
+    }
+  }
+
+  private CliFunctionResult executeWanCopyRegionFunctionInNewThread(
+      FunctionContext<Object[]> context,
+      Region<?, ?> region, String regionName, GatewaySender sender, long maxRate, int batchSize)
+      throws InterruptedException, ExecutionException, CancellationException {
+    String executionName = getExecutionName(regionName, sender.getId());
+    CompletableFuture<CliFunctionResult> future = null;
+    Executor executor = executorFactory.getExecutor((InternalCache) context.getCache());
+    try {
+      synchronized (executions) {
```

Comment:
Since `executions` would be moved to the service, I think you should only synchronize on that field over in the service. You could potentially pass a `Runnable` and a `String executionName` over to the service to be executed under synchronization. The service would then handle all use of the `executions` and probably the `executor`. Then the only thing this function would need is to invoke the method for executing tasks over in the service.

The code is then divided into two places. Everything that can be performed by multiple threads and multiple instances lives in `WanCopyRegionFunction`, while everything that must be one instance or fixed threads per Cache would live in `WanCopyRegionFunctionService`.

Reply:
An example of submitting `Runnables` to an execute method can be found in `AlertingAction` in geode-core. The purpose of `AlertingAction` is to surround the call to `run()` within a `ThreadLocal` but this could just as easily submit the `Runnable` to an `Executor` as well with related usage of your `executions` map.

Reply:
Nice change!
I have also split the `WanCopyRegionFunction` into two so that the new one `WanCopyRegionFunctionDelegate` is in charge of copying the entries.
The original one makes all the checks before invoking the new class and is also in charge of executing cancels via de new service provider.